bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/LedgersIndexRebuildOp.java

```diff
@@ -0,0 +1,214 @@
+ *   is running in read-only mode).
+ *   Losing the fenced status of a ledger is UNSAFE.
+ * - Sets the master key as an empty byte array. This is correct as empty master keys
+ *   are overwritten and we cannot use the password from metadata, and cannot know 100%
+ *   for sure how a digest for the password was generated.
+ */
+public class LedgersIndexRebuildOp {
+    private static final Logger LOG = LoggerFactory.getLogger(LedgersIndexRebuildOp.class);
+
+    private final ServerConfiguration conf;
+    private final boolean verbose;
+    private static final String LedgersSubPath = "ledgers";
+
+    public LedgersIndexRebuildOp(ServerConfiguration conf, boolean verbose) {
+        this.conf = conf;
+        this.verbose = verbose;
+    }
+
+    public boolean initiate()  {
+        LOG.info("Starting ledger index rebuilding");
+
+        String timestamp = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ").format(new Date());
+        String basePath = BookieImpl.getCurrentDirectory(conf.getLedgerDirs()[0]).toString();
+        String tempLedgersSubPath = LedgersSubPath + ".TEMP-" + timestamp;
+        Path tempPath = FileSystems.getDefault().getPath(basePath, tempLedgersSubPath);
+        Path currentPath = FileSystems.getDefault().getPath(basePath, LedgersSubPath);
+
+        LOG.info("Starting scan phase (scans journal and entry log files)");
+
+        try {
+            Set<Long> ledgers = new HashSet<>();
+            scanJournals(ledgers);
+            scanEntryLogFiles(ledgers);
+
+            LOG.info("Scan complete, found {} ledgers. "
+                    + "Starting to build a new ledgers index", ledgers.size());
+
+            KeyValueStorage newIndex = KeyValueStorageRocksDB.factory.newKeyValueStorage(
+                    basePath, tempLedgersSubPath, DbConfigType.Small, conf);
+            LOG.info("Created ledgers index at temp location {}", tempPath);
+
+            for (Long ledgerId : ledgers) {
+                DbLedgerStorageDataFormats.LedgerData ledgerData =
+                        DbLedgerStorageDataFormats.LedgerData.newBuilder()
+                                .setExists(true)
+                                .setFenced(true)
+                                .setMasterKey(ByteString.EMPTY).build();
+
+                byte[] ledgerArray = new byte[16];
+                ArrayUtil.setLong(ledgerArray, 0, ledgerId);
+                newIndex.put(ledgerArray, ledgerData.toByteArray());
+            }
+
+            newIndex.sync();
+            newIndex.close();
+        } catch (Throwable t) {
+            LOG.error("Error during rebuild, the original index remains unchanged", t);
+            delete(tempPath);
+            return false;
+        }
+
+        // replace the existing index
+        try {
+            Path prevPath = FileSystems.getDefault().getPath(basePath, LedgersSubPath + ".PREV-" + timestamp);
+            Files.move(currentPath, prevPath);
+            Files.move(tempPath, currentPath);
```

Comment:
If the first move fails, then the index should be in its original or back-up location. If the second move fails, the rebuilt index will still be in its temp location. Rather than try to auto fix after an error here, I'll add better logging to help the administrator fix it themselves. Seems like a highly unlikely scenario for a move to fail.